idaes/core/base/property_base.py

```diff
@@ -504,6 +505,47 @@ def report(self, index=(0), true_state=False, dof=False, ostream=None, prefix=""
 
         ostream.write("\n" + "=" * max_str_length + "\n")
 
+    def build_port(self, target_block, port_name, doc=None, subset=None):
+        """
+        Constructs a Port based on this StateBlock attached to the target block.
+
+        Args:
+            target_block - block to which Port should be attached
+            port_name - name to use for Port
+            doc - doc string or Prot object
+            subset - slicer representing a subset of indices of StateBlock which
+                should be included in Port
+
+        Returns:
+            Port object
+        """
+        if subset is None:
+            subset = self[...]
+
+        # Create empty Port
+        p = Port(doc=doc)
+        setattr(target_block, port_name, p)
```

Comment:
In principle the answer is probably no, but this would bring up a number of other questions. 

1. Where would the `References` used in the `Port` be placed? Strictly, I am not sure that they actually need to go anywhere, but my guess is it is probably best to attached them somewhere for future reference (i.e. someone reading the code, or someone who want to access those References programmatically).
2. If the `StateBlock` does not attach the `Port`, then that becomes the responsibility of something downstream (e.g. whatever called this method). This seems to be a bit of additional overhead and a chance for more confusion by users/readers. Note that you can always use `Port(extend=port_obj)` to make a copy of the `Port` elsewhere which I think would support most cases we could conceive.

Reply:
I think my preferred API would do something like return the both the port object and a list of references, with their names. Something like:
```python
    def build_port(self, port_name, doc=None, subset=None, index=None):
        """
        """
        if subset is None:
            subset = self[...]
        if index is None:
            index = self.index_set().first()

        # Create empty Port
        p = Port(doc=doc)

        # Get dict of Port members and names
        # Need to get a representative member of StateBlockDatas
        member_list = self[index].define_port_members()

        # Create References for port members
        name_and_reference_list = [] 
        for s in member_list:
            if not member_list[s].is_indexed():
                slicer = subset.component(member_list[s].local_name)
            else:
                slicer = subset.component(member_list[s].local_name)[...]

            r = Reference(slicer)
            s_name = "_" + s + "_" + port_name + "_ref"

            # Add Reference to Port
            p.add(r, s)
            name_and_reference_list.append((s_name, r))

        return p, name_and_reference_list
```

Reply:
Then the calling function would typically do something like:
```python
port, name_ref_list = state_block.build_port(self, name, doc)
self.add_component(name, port)
for ref_name, ref in name_ref_list:
    self.add_component(ref_name, ref)
```

Reply:
This does make the API more complicated, but I think this is worth it to not have a `setattr`/`add_component` embedded in some function where it is not obvious it would be happening. Here the `add_component` would happen in the `add_port` method, which I believe is much more clear.

Reply:
I see no problem with that in principle, but it would differ from how every other `build` method works. This would also require more extensive changes (I think) as this method is called by all the unit models, thus they would all need to be updated to do the `add_component`.

As such, I think this is a subject for a separate issue to discuss the benefits of this and where else it would need be applied for consistency.

Reply:
This shouldn't require extensive changes, as `build_port` doesn't currently exist, and if I'm remembering this correctly I'm not proposing any changes to `add_port`. But I see the point about consistency with other `build_` methods. Maybe `get_port_and_references` would be a better name for what I propose in the comment above.

I guess I just like, when possible, separating component "construction" from `add_component` calls. Mostly because then components can be added to whatever block we want, and also because embedded `add_component` calls have impeded my understanding of IDAES in the past.

Reply:
Actually, if I'm reading things correctly, `build_state_block` doesn't attach the constructed state block to any block, and just returns it. I think this was actually my inspiration for proposing this API for `build_port` (or `build_port_and_references`). So the proposed API is not completely inconsistent with other `build_` methods.

Reply:
Yep, that is what I'm proposing. `add_port` would call `build_port` (like it currently does), then attach the port and references to `self`. Agreed that this is more complicated than `build_state_block`, which is why a more descriptive name like `build_port_and_references` might be nice.